cmd/manager/main.go

```diff
@@ -93,10 +108,71 @@ func main() {
 		os.Exit(1)
 	}
 
+	if os.Getenv("ENABLE_WEBHOOKS") != "false" {
+		err = registerValidatingWebhook(&v1beta1.Instance{}, mgr)
+		if err != nil {
+			log.Error(err, "unable to create webhook")
+			os.Exit(1)
+		}
+	}
+
 	// Start the Cmd
 	log.Info("Starting the Cmd.")
 	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
 		log.Error(err, "unable to run the manager")
 		os.Exit(1)
 	}
 }
+
+// this is a fork of a code in controller-runtime to be able to pass in our own Validator interface
+// see kudo.Validator docs for more details\
+//
+// ideally in the future we should switch to just simply doing
+// err = ctrl.NewWebhookManagedBy(mgr).
+// For(&v1beta1.Instance{}).
+// Complete()
+//
+// that internally calls this method but using their own internal Validator type
+func registerValidatingWebhook(obj runtime.Object, mgr manager.Manager) error {
+	gvk, err := apiutil.GVKForObject(obj, mgr.GetScheme())
+	if err != nil {
+		return err
+	}
+	validator, isValidator := obj.(kudo.Validator)
+	if !isValidator {
+		log.Infof("skip registering a validating webhook, admission.Validator interface is not implemented", "GVK", gvk)
+		return nil
+	}
+	vwh := kudo.ValidatingWebhookFor(validator)
+	if vwh != nil {
+		path := generateValidatePath(gvk)
```

Comment:
I'm not sure we need the `generate` component.   I also didn't know path for what until diving into the code.   It's the API path.   Does `pathFor(gvk)` make sense?   or `apiPath(gvk)`? or `resourcePath(gvk)`?

Reply:
all these methods actually map to the exact same methods in controller-runtime. https://github.com/kubernetes-sigs/controller-runtime/blob/dc8357113a904bf02721efcde5d92937be39031c/pkg/builder/webhook.go#L163 as the names are not totally out of this world bad, I would probably stick with them, it can make it easier updating or migrating back to their implementation if they extend it in the future